File: packages/folio-core/domain/plan.go
package domain

import "time"

// PlanStatus tracks a plan through its life. A plan is the agent's stated
// intent for a piece of work; the todos under it are the steps.
type PlanStatus string

const (
	PlanDraft     PlanStatus = "draft"
	PlanActive    PlanStatus = "active"
	PlanDone      PlanStatus = "done"
	PlanAbandoned PlanStatus = "abandoned"
)

// IsTerminal reports whether the plan has reached an end state.
func (s PlanStatus) IsTerminal() bool {
	return s == PlanDone || s == PlanAbandoned
}

type Plan struct {
	ID        PlanID
	ProjectID ProjectID
	// TicketID is optional: an empty value means the plan sits directly under
	// the project rather than under one of its tickets.
	TicketID  TicketID
	Title     string
	Goal      string
	Status    PlanStatus
	Tags      []string
	CreatedBy UserID
	CreatedAt time.Time
	UpdatedAt time.Time

	// Progress is derived from the plan's todos on read, never persisted.
	Progress Progress
}

// Progress counts a plan's todos by completion, for clients that render a bar
// without fetching the whole list.
type Progress struct {
	Total int
	Done  int
}

// Percent returns completion 0-100; an empty plan reports 0.
func (p Progress) Percent() int {
	if p.Total == 0 {
		return 0
	}
	return p.Done * 100 / p.Total
}

File: apps/web/src/core/domain/plan.ts
import type { Branded } from './branded';
import type { IssueId } from './issue';
import type { ProjectId, UserId } from './project';

export type PlanId = Branded<string, 'PlanId'>

export const planId = (value: string): PlanId => value as PlanId

export const PLAN_STATUS = {
	DRAFT: 'draft',
	ACTIVE: 'active',
	DONE: 'done',
	ABANDONED: 'abandoned',
} as const

export const PLAN_STATUSES = [PLAN_STATUS.DRAFT, PLAN_STATUS.ACTIVE, PLAN_STATUS.DONE, PLAN_STATUS.ABANDONED] as const

export type PlanStatus = (typeof PLAN_STATUSES)[number]

export const DEFAULT_PLAN_STATUSES: readonly PlanStatus[] = PLAN_STATUSES.filter(status => status !== PLAN_STATUS.DONE && status !== PLAN_STATUS.ABANDONED)

export type Plan = {
	readonly id: PlanId
	readonly projectId: ProjectId
	readonly ticketId: IssueId | undefined
	readonly title: string
	readonly goal: string
	readonly status: PlanStatus
	readonly tags: readonly string[]
	readonly createdBy: UserId | undefined
	readonly createdAt: Date
	readonly updatedAt: Date
}

export const isTerminal = (status: PlanStatus): boolean => status === PLAN_STATUS.DONE || status === PLAN_STATUS.ABANDONED
